vm-virtio/src/balloon.rs

```diff
@@ -0,0 +1,569 @@
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(PAUSE_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+
+        const EPOLL_EVENTS_LEN: usize = 100;
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); EPOLL_EVENTS_LEN];
+
+        'epoll: loop {
+            let num_events = match epoll::wait(epoll_fd, -1, &mut events[..]) {
+                Ok(res) => res,
+                Err(e) => {
+                    if e.kind() == io::ErrorKind::Interrupted {
+                        // It's well defined from the epoll_wait() syscall
+                        // documentation that the epoll loop can be interrupted
+                        // before any of the requested events occurred or the
+                        // timeout expired. In both those cases, epoll_wait()
+                        // returns an error of type EINTR, but this should not
+                        // be considered as a regular error. Instead it is more
+                        // appropriate to retry, by calling into epoll_wait().
+                        continue;
+                    }
+                    return Err(DeviceError::EpollWait(e));
+                }
+            };
+
+            for event in events.iter().take(num_events) {
+                let ev_type = event.data as u16;
+
+                match ev_type {
+                    RESIZE_EVENT => {
+                        if let Err(e) = self.resize.evt.read() {
+                            error!("Failed to get resize event: {:?}", e);
+                            break 'epoll;
+                        } else {
+                            let mut need_break = false;
+                            let r = {
+                                let mut config = self.config.lock().unwrap();
+                                config.num_pages = (self.resize.get_size() >> PAGE_SHIFT) as u32;
+                                if let Err(e) = self.signal(&VirtioInterruptType::Config, None) {
+                                    need_break = true;
+                                    Err(Error::ResizeTriggerFail(e))
+                                } else {
+                                    Ok(())
+                                }
+                            };
+                            if let Err(e) = &r {
+                                error!("{:?}", e);
+                            }
+                            self.resize.send(r);
+                            if need_break {
+                                break 'epoll;
+                            }
+                        }
+                    }
+                    INFLATE_QUEUE_AVAIL_EVENT => {
+                        if let Err(e) = self.inflate_queue_evt.read() {
+                            error!("Failed to get inflate queue event: {:?}", e);
+                            break 'epoll;
+                        } else if let Err(e) = self.process_queue(ev_type) {
```

Comment:
If you don't mind, I would like to keep it because this PR only implements the basic functions of the balloon.  There are other vq need to be handled by this function.

Reply:
Ok fair enough :)